eodatasets3/images.py

```diff
@@ -559,157 +562,174 @@ def write_from_ndarray(
+        *input_geobox.bounds,
+    )
+    reproj_grid = GridSpec(
+        (reprojected_height, reprojected_width), reprojected_transform, crs=out_crs
+    )
+    # Calculate combined nodata mask
+    nulls = numpy.zeros(input_geobox.shape, dtype="bool")
+    for band_no, band_path in enumerate(rgb, start=1):
+        with rasterio.open(band_path) as ds:
+            if ds.count != 1:
+                raise NotImplementedError(
+                    "multi-band measurement files aren't yet supported"
                 )
-                scaled = scaled.astype("uint8")
-                scaled[nulls] = 0
+            nulls |= ds.read(1) == ds.nodata
+
+    ql_write_args = dict(
+        driver="GTiff",
+        dtype="uint8",
+        count=len(rgb),
+        width=reproj_grid.shape[1],
+        height=reproj_grid.shape[0],
+        transform=reproj_grid.transform,
+        crs=reproj_grid.crs,
+        nodata=0,
+        tiled="yes",
+    )
+    # Only set blocksize on larger imagery; enables reduced resolution processing
+    if reproj_grid.shape[0] > 512:
+        ql_write_args["blockysize"] = 512
+    if reproj_grid.shape[1] > 512:
+        ql_write_args["blockxsize"] = 512
+
+    with rasterio.open(dest_path, "w", **ql_write_args) as ql_ds:
+        ql_ds: DatasetWriter
+        for band_no, band_path in enumerate(rgb, start=1):
+            with rasterio.open(band_path) as ds:
+                ds: DatasetReader
+                rescaled_data = rescale_intensity(
+                    ds.read(1), in_range=src_range, out_range=(1, 255)
+                )
+                rescaled_data[nulls] = 0
+
+                reprojected_data = numpy.zeros(reproj_grid.shape, dtype=numpy.uint8)
+                reproject(
+                    rescaled_data,
+                    reprojected_data,
+                    src_crs=ds.crs,
+                    src_transform=ds.transform,
+                    src_nodata=0,
+                    dst_crs=reproj_grid.crs,
+                    dst_nodata=0,
+                    dst_transform=reproj_grid.transform,
+                    resampling=resampling,
+                    num_threads=2,
+                )
+                del rescaled_data
+                ql_ds.write(reprojected_data, band_no)
+                del reprojected_data
+    del nulls
+    return reproj_grid
+
+
+def rescale_intensity(
+    image: np.ndarray,
+    in_range: Tuple[int, int],
+    out_range: Tuple[int, int],
+    dtype=np.uint8,
+) -> np.ndarray:
+    """
+    Based on scikit-image's rescale_intensity, but does fewer copies/allocations of the array.
 
-                out_ds.write(scaled, band)
+    (and it saves us bringing in the entire dependency for one small method)
+    """
+    imin, imax = in_range
+    omin, omax = out_range
+
+    np.clip(image, imin, imax, out=image)
+    image -= imin
+    # We allocate a float array here, as with the original function.
+    # (I'm not sure how bad int truncation would affect us?)
```

Comment:
I've tested out_ranges of [1, 255] and [0, 255] and they don't have any truncation side effects.  But my interests here are purely for display, as such the range [0, 255] or anything in-between is all I'm caring about.  Anything outside this range shouldn't have an issue either, as it would be following the same stretch logic.  Or we could include a set check of the range(0, 255) to ensure an input exists only within that range.